lib/registry/src/lib.rs

```diff
@@ -1180,6 +1189,176 @@ pub fn get_all_available_registries() -> Result<Vec<String>, String> {
+    }
+
+    Ok(())
+}
+
+/// Returns a list of all installed webc packages
+pub fn get_all_installed_webc_packages() -> Vec<RemoteWebcInfo> {
+    let dir = match get_webc_dir() {
+        Some(s) => s,
+        None => return Vec::new(),
+    };
+
+    let read_dir = match std::fs::read_dir(dir) {
+        Ok(s) => s,
+        Err(_) => return Vec::new(),
+    };
+
+    read_dir
+        .filter_map(|r| Some(r.ok()?.path()))
+        .filter_map(|path| {
+            webc::WebCMmap::parse(
+                path,
+                &webc::ParseOptions {
+                    parse_atoms: false,
+                    parse_volumes: false,
+                    ..Default::default()
+                },
+            )
+            .ok()
+        })
+        .filter_map(|webc| {
+            let mut checksum = webc.checksum.as_ref().map(|s| &s.data)?.to_vec();
+            while checksum.last().copied() == Some(0) {
+                checksum.pop();
+            }
+            let hex_string = hex::encode(&checksum);
+            Some(RemoteWebcInfo {
+                checksum: hex_string,
+                manifest: webc.manifest.clone(),
+            })
+        })
+        .collect()
+}
+
+/// Returns the checksum of the .webc file, so that we can check whether the
+/// file is already installed before downloading it
+pub fn get_remote_webc_checksum(url: &Url) -> Result<String, String> {
+    let request_max_bytes = webc::WebC::get_signature_offset_start() + 4 + 1024 + 8 + 8;
+    let data = get_webc_bytes(url, Some(0..request_max_bytes))?;
+    let mut checksum = webc::WebC::get_checksum_bytes(&data)
+        .map_err(|e| format!("{e}"))?
+        .to_vec();
+    while checksum.last().copied() == Some(0) {
+        checksum.pop();
+    }
+    let hex_string = hex::encode(&checksum);
+    Ok(hex_string)
+}
+
+/// Before fetching the entire file from a remote URL, just fetch the manifest
+/// so we can see if the package has already been installed
+pub fn get_remote_webc_manifest(url: &Url) -> Result<RemoteWebcInfo, String> {
+    // Request up unti manifest size / manifest len
+    let request_max_bytes = webc::WebC::get_signature_offset_start() + 4 + 1024 + 8 + 8;
+    let data = get_webc_bytes(url, Some(0..request_max_bytes))?;
+    let mut checksum = webc::WebC::get_checksum_bytes(&data)
+        .map_err(|e| format!("{e}"))?
+        .to_vec();
+    while checksum.last().copied() == Some(0) {
+        checksum.pop();
+    }
+    let hex_string = hex::encode(&checksum);
+
+    let (manifest_start, manifest_len) =
+        webc::WebC::get_manifest_offset_size(&data).map_err(|e| format!("{e}"))?;
+    let data_with_manifest = get_webc_bytes(url, Some(0..manifest_start + manifest_len))?;
+    let manifest = webc::WebC::get_manifest(&data_with_manifest).map_err(|e| e.to_string())?;
+    Ok(RemoteWebcInfo {
+        checksum: hex_string,
+        manifest,
+    })
+}
+
+fn setup_webc_client(url: &Url) -> Result<reqwest::blocking::RequestBuilder, String> {
```

Comment:
Should this return a proper error type instead of a `String`?

Reply:
I tried to use `String` in the `registry` crate to avoid a dependency on `anyhow`, but maybe it's just better to use anyhow here.

I am against strongly typed errors (because they add a lot of maintenance work) - unless the caller can recover from the error. For the use in the CLI, the caller cannot recover, but I can use `anyhow`, with `context()`.